### PR TITLE
Update/Close race

### PR DESCRIPTION
**Pull request by andrewgordstewart**
- [channel-provider] Make some type-related changes to the channel-provider package to make it possible/more comfortable to catch wallet errors
- [web3torrent] Retry `CloseChannel` when it fails due to a NotYourTurn error
- [web3torrent] Catch NotYourTurn errors when calling `UpdateChannel`

### CODE REVIEW

File: packages/web3torrent/src/clients/payment-channel-client.ts [hunk 1]
```diff
@@ -280,31 +282,31 @@ export class PaymentChannelClient {
...
-     * Now, any UpdateChannel RPC should be finished, so when it's my turn, the CloseChannel
-     * call should succeed.
-     */
-
     const closing = this.channelState(channelId)
-      .pipe(first(cs => this.canUpdateChannel(cs)))
-      .subscribe(cs => {
+      .pipe(
+        filter(cs => this.canUpdateChannel(cs)),
+        take(5)
```
Comment (snario) [inline packages/web3torrent/src/clients/payment-channel-client.ts:288]: Might warrant a comment as to why 5 is here

File: packages/web3torrent/src/clients/payment-channel-client.ts [hunk 2]
```diff
@@ -390,10 +392,18 @@ export class PaymentChannelClient {
...
+    try {
+      await this.updateChannel(channelId, {
+        beneficiary: {...beneficiary, balance: add(beneficiary.balance, amountWillPay)},
+        payer: {...payer, balance: subtract(payer.balance, amountWillPay)}
+      });
+    } catch (error) {
+      if (error.error.code === ErrorCode.UpdateChannel.NotYourTurn) {
+        logger.warn({channelId}, 'Possible race condition detected');
+      } else {
+        logger.error({error}, 'makePamyent: Unexpected error');
```
Comment (snario) [inline packages/web3torrent/src/clients/payment-channel-client.ts:404]:
```suggestion
        logger.error({error}, 'makePayment: Unexpected error');
```

File: packages/web3torrent/src/clients/payment-channel-client.ts [hunk 3]
```diff
@@ -280,31 +282,31 @@ export class PaymentChannelClient {
...
+        take(5) // Limit to 5 attempts
+      )
+      .subscribe(async cs => {
         logger.info(
           {channelId, cs, me: this.mySigningAddress},
           "It's my turn, closing the channel"
         );
-        this.channelClient.closeChannel(channelId);
+        try {
+          await this.channelClient.closeChannel(channelId);
```
Comment (kerzhner) [inline packages/web3torrent/src/clients/payment-channel-client.ts:296]: In the case when the seeder is initiating the channel close, this call always race (payment acceptance) [https://github.com/statechannels/monorepo/blob/master/packages/web3torrent/src/library/web3torrent-lib.ts#L340]

File: .circleci/config.yml [hunk 4]
```diff
@@ -384,9 +384,9 @@ jobs:
       - store_artifacts:
           path: /tmp/ganache.log
       - store_artifacts:
-          path: /tmp/seed-download-cancel.pino.log
+          path: /tmp/multiple-leecher.pino.log
       - store_artifacts:
-          path: /tmp/seed-download-cancel.browser.log
+          path: /tmp/multiple-leecher.browser.log
```
[the remark this answers is not in the document]
Comment (andrewgordstewart) [reply .circleci/config.yml:389]: This is currently in process in the `ags/fix-artifacts` branch

File: packages/channel-client/src/types.ts [hunk 5]
```diff
@@ -67,3 +68,19 @@ export interface EventsWithArgs {
   // TODO: Is `ChannelResult` the right type to use here?
   ChannelProposed: [ChannelResult];
 }
+
+export const ErrorCode: ErrorCodes = {
+  EnableEthereum: {EthereumNotEnabled: 100},
```
[the remark this answers is not in the document]
Comment (andrewgordstewart) [reply packages/channel-client/src/types.ts:73]: They are fairly arbitrary. When organizing the client schema last week, I opted to use 400's for one API, 300's for another, etc.

 The code added here is simply a constant value -- the codes were already in place in the `ErrorCodes` type.

File: packages/web3torrent/src/clients/payment-channel-client.ts [hunk 6]
```diff
@@ -280,31 +282,31 @@ export class PaymentChannelClient {
...
-     */
-
     const closing = this.channelState(channelId)
-      .pipe(first(cs => this.canUpdateChannel(cs)))
-      .subscribe(cs => {
+      .pipe(
+        filter(cs => this.canUpdateChannel(cs)),
+        take(5) // Limit to 5 attempts
+      )
+      .subscribe(async cs => {
```
[the remark this answers is not in the document]
Comment (andrewgordstewart) [reply packages/web3torrent/src/clients/payment-channel-client.ts:290]: I think so. I was thinking of adding it but never got to it

File: packages/web3torrent/src/clients/payment-channel-client.ts [hunk 7]
```diff
@@ -280,31 +282,31 @@ export class PaymentChannelClient {
...
+        try {
+          await this.channelClient.closeChannel(channelId);
+        } catch (error) {
+          if (error.error.code === ErrorCode.CloseChannel.NotYourTurn) {
+            // Perhaps the application calls UpdateChannel around the same time as the CloseChannel call
+            logger.warn(
+              {channelId},
+              'Possible race condition detected -- is a concurrent UpdateChannel occurring?'
+            );
+            return;
```
[the remark this answers is not in the document]
Comment (andrewgordstewart) [reply packages/web3torrent/src/clients/payment-channel-client.ts:304]: I don't think it's needed but I view it as being explicit that we are ok catching this error

File: packages/web3torrent/src/clients/payment-channel-client.ts (hunk 3 above)
Comment (andrewgordstewart) [reply packages/web3torrent/src/clients/payment-channel-client.ts:296]: So, the torrent client should probably switch modes and not accept payments. Then, there would be at most once race.

Comment (andrewgordstewart) [reply packages/web3torrent/src/clients/payment-channel-client.ts:296]: So, the torrent client should probably switch modes and not accept payments. Then, there would be at most once race.

**Pull request merged**